resources/Microsoft.Windows.Setting.WindowsUpdate/Microsoft.Windows.Setting.WindowsUpdate.psm1

```diff
@@ -0,0 +1,377 @@
+        $typeInfo = $value.GetType().Name
+
+        if ($typeInfo -eq 'Boolean')
+        {
+            $value = [int]$value
+        }
+
+        # validate the value of UserChoiceActiveHoursEnd and UserChoiceActiveHoursStart to be between 0 and 24
+        Assert-UserChoiceValue -KeyName $key -Value $value
+
+        # validate the value of DownloadRateBackgroundPct, DownloadRateForegroundPct and UpRatePctBandwith to be between 0 and 100
+        Assert-RatePercentageValue -KeyName $key -Value $value
+
+        # validate the value of UpRatePctBandwith to be between 5 and 500
+        Assert-UpRateValue -KeyName $key -Value $value
+
+        if (-not (DoesRegistryKeyPropertyExist -Path $Path -Name $key))
+        {
+            $null = New-ItemProperty -Path $Path -Name $key -Value $value -PropertyType 'DWord' -Force  
+        }
+
+        Write-Verbose -Message "Setting $key to $($RegistryKeyProperty[$key])"
+        Set-ItemProperty -Path $Path -Name $key -Value $value
+    }
+} 
+
+function Assert-UpRateValue 
+{
+    param (
+        [Parameter(Mandatory)]
+        [string] $KeyName,
+
+        [Parameter(Mandatory)]
+        [int] $Value
+    )
+
+    if ($KeyName -eq 'UpRatePctBandwidth' -and $Value -notin (5..500))
+    {
+        Throw "You are specifying a percentage value, which must be between 5 and 500. The value you provided is $Value. Please provide a value between 5 and 500."
+    }
+}
+
+function Assert-RatePercentageValue
+{
+    param (
+        [Parameter(Mandatory)]
+        [string] $KeyName,
+
+        [Parameter(Mandatory)]
+        [int] $Value
+    )
+
+    if ($KeyName -in ('DownloadRateBackgroundPct', 'DownloadRateForegroundPct', 'UpRatePctBandwidth') -and $Value -notin (0..100))
+    {
+        # TODO: It might be beneficial to add `Reasons` and not throw, only return statement
+        Throw "You are specifying a percentage value, which must be between 0 and 100. The value you provided is $Value. Please provide a value between 0 and 100."
+    }
+}
+
+function Assert-UserChoiceValue 
+{
+    param (
+        [Parameter(Mandatory)]
+        [string] $KeyName,
+
+        [Parameter(Mandatory)]
+        [int] $Value
+    )
+
+    if ($KeyName -in ('UserChoiceActiveHoursEnd', 'UserChoiceActiveHoursStart') -and $Value -notin (0..24))
+    {
+        Throw "Value must be between 0 and 24"
+    }
+}
+
+function Assert-DownloadRate
+{
+    param (
+        [Parameter(Mandatory)]
+        [hashtable] $Parameters
+    )
+
+    if ($Parameters.ContainsKey('DownloadRateBackgroundPct') -or $Parameters.ContainsKey('DownloadRateForegroundPct'))
+    {
+        if ($Parameters.ContainsKey('DownloadRateBackgroundBps') -or $Parameters.ContainsKey('DownloadRateForegroundBps'))
+        {
+            Throw "Cannot set both DownloadRateBackgroundPct/DownloadRateForegroundPct and DownloadRateBackgroundBps/DownloadRateForegroundBps"
+        }
+    }
+}
+
+function Initialize-WindowsUpdate 
```

Comment:
It feels like this function shouldn’t be exported. I suggest either updating the psd1 to not export it, or somehow moving it to a non-exported namespace. 

In fact, most of the functions don’t seem like they need to be exported, and could even potentially live inside the class as static functions

Reply:
Mhmmm. I might have to adjust this in the `New-DscResourceModule`. I have overlooked this one. Nothing has to be exported, expect the `DscResourcesToExport`. I'll update them all.